community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/input/InputEntityReader.java

```diff
@@ -0,0 +1,186 @@
+import org.neo4j.helpers.collection.PrefetchingIterator;
+import org.neo4j.io.fs.StoreChannel;
+import org.neo4j.kernel.impl.transaction.log.LogPositionMarker;
+import org.neo4j.kernel.impl.transaction.log.PhysicalLogVersionedStoreChannel;
+import org.neo4j.kernel.impl.transaction.log.ReadAheadLogChannel;
+import org.neo4j.kernel.impl.transaction.log.ReadPastEndException;
+import org.neo4j.kernel.impl.transaction.log.ReadableLogChannel;
+import org.neo4j.unsafe.impl.batchimport.InputIterator;
+
+import static org.neo4j.helpers.Format.KB;
+import static org.neo4j.kernel.impl.transaction.log.LogVersionBridge.NO_MORE_CHANNELS;
+
+/**
+ * Abstract class for reading cached entities previously stored using {@link InputEntityCacher} or derivative.
+ */
+abstract class InputEntityReader<ENTITY extends InputEntity> extends PrefetchingIterator<ENTITY>
+        implements InputIterator<ENTITY>
+{
+    protected final ReadableLogChannel channel;
+    private final LogPositionMarker positionMarker = new LogPositionMarker();
+    private int lineNumber;
+    private final Group[] previousGroups;
+    private final PrimitiveIntObjectMap<String> tokens = Primitive.intObjectMap();
+
+    InputEntityReader( StoreChannel channel, StoreChannel header, int bufferSize, int groupSlots ) throws IOException
+    {
+        this.previousGroups = new Group[groupSlots];
+        for ( int i = 0; i < groupSlots; i++ )
+        {
+            previousGroups[i] = Group.GLOBAL;
+        }
+        this.channel = reader( channel, bufferSize );
+        readHeader( header );
+    }
+
+    private ReadAheadLogChannel reader( StoreChannel channel, int bufferSize ) throws IOException
+    {
+        return new ReadAheadLogChannel(
+                new PhysicalLogVersionedStoreChannel( channel, 0, (byte) 0 ), NO_MORE_CHANNELS, bufferSize );
+    }
+
+    private void readHeader( StoreChannel header ) throws IOException
+    {
+        try ( ReadableLogChannel reader = reader( header, 8*KB ) )
+        {
+            for ( short id = 0; reader.get() == 1; id++ )
+            {
+                tokens.put( id, (String) ValueType.stringType().read( reader ) );
+            }
+        }
+    }
+
+    @Override
+    protected final ENTITY fetchNextOrNull()
+    {
+        long position = position();
+        try
+        {
+            lineNumber++;
+            return readNextOrNull();
+        }
+        catch ( ReadPastEndException e )
+        {
+            return null;
+        }
+        catch ( IOException e )
+        {
+            throw new InputException( "Couldn't read cached node data", e );
+        }
+        catch ( Throwable t )
+        {
+            System.out.println( "at line " + lineNumber + " and position " + position );
+            throw t;
+        }
+    }
+
+    protected abstract ENTITY readNextOrNull() throws IOException;
+
+    protected Object properties() throws IOException
+    {
+        short count = channel.getShort();
```

Comment:
right, although the format of the header makes it so that `ReadPastEndException` shouldn't need to be thrown, so if that exception indeed should be thrown from `readHeader` then it should get propagated out. The data stream doesn't have such format, I'd guess mostly to save 1 byte per entry. Oh, actually I can change the format so that the end is predictable w/o any overhead. I'll do that instead so that it gets clearer.